core/src/main/scala/org/http4s/headers/Forwarded.scala

```diff
@@ -193,7 +268,80 @@ object Forwarded
       ParseFailure("missing host", s"no host defined in the URI '$uri'")
   }
 
-  override def parse(s: String): ParseResult[Forwarded] = parser.HttpHeaderParser.FORWARDED(s)
+  override def parse(s: String): ParseResult[Forwarded] =
+    parser.parseAll(s).left.map { e =>
+      ParseFailure("Invalid header", e.toString)
+    }
+
+  private val parser: P[Forwarded] = {
+    // https://tools.ietf.org/html/rfc7239#section-4
+
+    // A utility so that we can decode multiple pairs and join them in a single element
+    trait Pair {
+      def create: Element
+      def merge(e: Element): Element
+    }
+
+    object Pair {
+      def apply(c: Element, m: Element => Element): Pair = new Pair {
+        override def create: Element = c
+        override def merge(e: Element): Element = m(e)
+      }
+    }
+
+    def quoted[A](p: P[A]): P[A] =
+      Rfc7230.token
+        .orElse(Rfc7230.quotedString)
+        .flatMap(str =>
```

Comment:
Do we have to flatMap these, or is each `p` sufficient on its own?  Is it the quoting that makes this so complicated?

Reply:
I looked at the previous parser and this was more or less what it was doing. I imagined `quoted-string` was more than putting the quotes in the beginning and the end but it isn't, so I think I can rewrite this as 
`oneOf(p.between('"', '"'), p)`.

   char('"') *> p <

Reply:
:point_up: this is not true, i'm now getting more failures because the host parser is consuming the text of the following pairs, e.g. `host="http4s.org;for=_kadabra;proto=http;by=_abra"`. Maybe `quoted-string` is more restrictive than `Host`? Will investigate later.